Add cache for application privileges

ywangd opened the pull request:
Add caching support for application privileges to reduce number of round-trips to security index when building application privilege descriptors.

A few key points of the changes:
* Per discussion #54317, the main cache is keyed by concrete application name and values are a set of application privilege descriptors.
* A secondary cache is also added to map a set of application expressions (i.e. with wildcard) to a set of concrete application names.
* Due to the cache design, privilege retrieving in `NativePrivilegeStore` is changed to always fetching all privilege documents for a given application.
* The change is applied to all places including "get privilege", "has privileges" API and  `CompositeRolesStore` (for authentication). 
  - Initially I wanted to exclude `CompositeRolesStore` from the caching. But this means no code can be deleted from `NativePrivilegeStore`. We basically have to add "retriving by application name" on top of existing query logic. For simplicity, I later decided to include it so that the query part can be largely simplified.
* Added an API to clear privilege cache, which is also used internally when adding/deleting privileges. These operations already clear role cache, the privilege cache invalidation is added on top of it.
* Add security index state listener to clear cache on state changes.
* Docs added
* Yaml tests added

Resolves: #54317

ywangd committed:
WIP

ywangd committed:
Add tests

ywangd committed:
Add invalidation

ywangd committed:
Add some corrurrency liveness protection

ywangd committed:
Add API for both rest and transport

ywangd committed:
Simplify NativePrivilegeStore

ywangd committed:
Minor

ywangd committed:
More tweak

ywangd committed:
Add single node tests

Rest tests
Documents
Other tweaks and optimizations

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authorization)

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -83,47 +100,120 @@
+                    final Set<String> allApplicationNames =
+                        Stream.concat(cacheStatus.v2().keySet().stream(), mapOfFetchedDescriptors.keySet().stream())
+                            .collect(Collectors.toUnmodifiableSet());
+                    // Avoid caching potential stale results.
+                    // TODO: It is still possible that cache gets invalidated immediately after the if check
+                    if (invalidationCounter == numInvalidation.get()) {
```
This pattern is used in `CompositeRolesStore`. However it is still possible that the cache gets invalidated immediately after the `if` check. The window for it to happen is much small, but still exists in theory.

The other pattern is to use `ListenableFuture`. This would solves the stale entry problem here because the `future` itself is removed from the cached. So adding items to the removed `future` has no impact to the cache. However, this pattern could potentially have a deadlock issue? If the thread computing for future crashes, will all the other thread waiting for it get stuck? If this is true, I'd prefer to use the first pattern since a stale entry (with very low chance) is less harmful than deadlocking.

tvernum replied:
I think a ReadWriteLock would solve the problem. Treat the invalidator as the writer and the cache population as the reader. The invalidator would need exclusive access, but we could support multiple populators.

I'll need to think about it again when it's not midnight, but I think it's reasonable (if used alongside the `invalidationCounter` so the lock window is small).

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -83,47 +100,120 @@
+                    // TODO: It is still possible that cache gets invalidated immediately after the if check
+                    if (invalidationCounter == numInvalidation.get()) {
+                        // Do not cache the names if expansion has no effect
+                        if (allApplicationNames.equals(applicationNamesCacheKey) == false) {
```
This is a minor optimisation so we do not store an etnry of `{Set('app-1'), Set('app-1')}`, where there is no expansion at all.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -133,17 +223,69 @@ public void getPrivileges(Collection<String> applications, Collection<String> na
+        logger.debug("Invalidating application privileges caches for: {}", updatedApplicationNames);
+        numInvalidation.incrementAndGet();
+        // Always completely invalidate application names cache due to wildcard
+        applicationNamesCache.invalidateAll();
+        final Set<String> uniqueNames = Set.copyOf(updatedApplicationNames);
```
For invalidation, the application names cache is always completey invalidated since it requires some effort to identify the applicable entries. We could do this, but the gain may not be much.

tvernum replied:
Remind me, what's Kibana's typical usage pattern for querying? Does it use wildcards for the application name?

If so, I think invalidating the name cache means that invalidating for a single application (which might not even exist) would effectively invalidate the whole cache because it would mean that querying for `kibana*` would end up not using any cache at all.

ywangd replied:
Kibana always sends a single concrete application name `kibana-.kibana`. So it should be fine for Kibana's typical usage. 

But let me know if you think it is still necessary. The logic would look like something as the follows:
```java
foreach cache key (type is Set<String>)
   foreach key member
     foreach application
        if (key member == application) or (key member is a wildcard and matches application)
          invalidate the cache key
```

Java code would be
```java
StreamSupport.stream(applicationNamesCache.keys().spliterator(), false)
    .filter(keys -> keys.contains("*")
        || Sets.intersection(keys, uniqueNames).isEmpty() == false
        || keys.stream().filter(k -> k.endsWith("*")).anyMatch(
            k -> uniqueNames.stream().anyMatch(n -> n.regionMatches(false, 0, k, 0, k.length()-1))))
    .forEach(applicationNamesCache::invalidate);

tvernum replied:
Let's leave it - realistically we're talking about Kibana only, so clearing all applications isn't actually going to hurt anyone, and the only time _we_ will clear the cache is on privilege update which happens when you install a new Kibana version.

ywangd committed:
checkstyle

ywangd committed:
More checkstyle

ywangd committed:
checkstyle again

ywangd committed:
Fix json API file

ywangd committed:
Update API json file

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/privilege/TransportClearPrivilegesCacheAction.java:
```diff
@@ -0,0 +1,82 @@
+        } else {
+            privilegesStore.invalidate(Arrays.asList(request.getApplicationNames()));
+        }
+        rolesStore.invalidateAll();
```
I'm not sure about this. It seems like this API ends up doing something other than what it was supposed to, just because we assume that the caller wants it.
I understand why - if the privileges have changed then the roles cache is probably wrong, but it seems like it's chain side-effects together.

ywangd replied:
From just the API point of view, you are right that these two should not be tied together. There are valid use cases when user only wants to actively clear privileges cache. I did this because the two are always tied together in `NativePrivilegesStore` since it was clearing role cache before my change.

I tried to avoid nested callbacks (clear role cache then clear privileges cache). It seems OK and more efficient by just looking at `NativePrivilegesStore`. But it does feel wrong from pure API side.

I could either just go with nested callback or create a transport layer only action to clear both caches. So it is not exposed at REST layer and still has the efficiency on transport layer. But this does lead some code redundancy. Another option is to have a query parameter for the clear privilege cache API. When set to true, it clears both caches. What do you think?

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/privilege/TransportClearPrivilegesCacheAction.java:
```diff
@@ -0,0 +1,82 @@
+        if (request.getApplicationNames() == null || request.getApplicationNames().length == 0) {
+            privilegesStore.invalidateAll();
+        } else {
+            privilegesStore.invalidate(Arrays.asList(request.getApplicationNames()));
```
```suggestion
            privilegesStore.invalidate(List.of(request.getApplicationNames()));
```

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/ClearPrivilegesCacheResponse.java:
```diff
@@ -0,0 +1,64 @@
+import java.io.IOException;
+import java.util.List;
+
+public class ClearPrivilegesCacheResponse extends BaseNodesResponse<ClearPrivilegesCacheResponse.Node>
```
Separate to this PR, it feels like we could consolidate these duplicate classes into a common base class.

ywangd replied:
A common base class for all `ClearXxxCacheResponse`?

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/CacheIteratorHelper.java:
```diff
@@ -56,4 +56,17 @@ public void removeKeysIf(Predicate<K> removeIf) {
         }
     }
+
+    public void removeValuesIf(Predicate<V> removeIf) {
```
Where do we call this? I think we should avoid having this method if we can.

ywangd replied:
This is indeed not needed anymore. Will remove.

ywangd commented:
Resolves #54317

ywangd committed:
WIP

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

Co-Authored-By: Tim Vernum <tim@adjective.org>

ywangd committed:
Merge branch 'es-54317-app-privilege-cache' of github.com:ywangd/elasticsearch into es-54317-app-privilege-cache

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/privilege/TransportClearPrivilegesCacheAction.java

Co-Authored-By: Tim Vernum <tim@adjective.org>

ywangd committed:
Merge branch 'es-54317-app-privilege-cache' of github.com:ywangd/elasticsearch into es-54317-app-privilege-cache

ywangd committed:
WIP

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -83,47 +100,120 @@
+            applicationNamesCacheBuilder.weigher((k, v) -> k.size() + v.size());
+        }
+        applicationNamesCache = applicationNamesCacheBuilder.build();
+        applicationNamesCacheHelper = new CacheIteratorHelper<>(applicationNamesCache);
```
Right I don't need interate for either of the two caches. I'll remove both helpers.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -83,47 +100,120 @@
+        if (concreteApplicationNames != null && concreteApplicationNames.size() == 0) {
+            logger.debug("returning empty application privileges as application names result in empty list");
+            listener.onResponse(Collections.emptySet());
+
```
Ack

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -83,47 +100,120 @@
+            listener.onResponse(Collections.emptySet());
+
+        } else {
+            final Tuple<Set<String>, Map<String, Set<ApplicationPrivilegeDescriptor>>> cacheStatus;
```
This complexity is due to an attempt to optimize number of documents to be fetched from index. It can be simplified if we always fetching everything when hitting the index is unavoidable (as you suggested below).

ywangd replied:
Once we remove this optimization the `Tuple<Set<String>, Map<String, Set<ApplicationPrivilegeDescriptor>>>` data structure is no longer necessary. So complexity will definitely be reduced

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -83,47 +100,120 @@
+                final long invalidationCounter = numInvalidation.get();
+                // Always fetch all privileges of an application for caching purpose
+                logger.debug("Fetching application privilege documents for: {}", cacheStatus.v1());
+                innerGetPrivileges(cacheStatus.v1(), ActionListener.wrap(fetchedDescriptors -> {
```
I was trying to optimize and reduce the number of documents to fetch. But maybe it is not worth the effort.

ywangd committed:
Address feedback

ywangd committed:
Checkstyle

ywangd committed:
Address feedback

ywangd committed:
Skip yaml tests for 7.x

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

ywangd committed:
Fix yaml tests

ywangd commented:
Updated to address feedbacks, also added yaml tests and docs, removed Draft status from the PR. Here is a list of more important changes:

* Simplified caching reuse mechanism so that if we need to fetch some privileges from index, we will always fetch for all requested applications (and cache them afterwards). So the response is either all from cache or all from index search, never a mix of both.
* Add a `clearRolesCache` parameter to `ClearPrivilegesCacheRequest`. When it is set, the roles cache will be cleared after privileges cahce. This parameter is only used internally at transport layer when put/delete privileges. It is Not exposed on rest layer.

ywangd committed:
Add HLRC relevant files

ywangd commented:
* Added relevant HLRC classes and docs for the new invalidate privileges cache API.

ywangd commented on x-pack/plugin/src/test/resources/rest-api-spec/api/security.clear_cached_privileges.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "security.clear_cached_privileges":{
+    "documentation":{
+      "url":"https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-clear-role-cache.html",
```
Absolutely. This was a placeholder. It's updated now that the actual doc is in place.

ywangd committed:
check style

ywangd committed:
Fix broken docs

ywangd committed:
Add bwc for the new transport action

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/privilege/TransportClearPrivilegesCacheAction.java:
```diff
@@ -0,0 +1,96 @@
+        assert request.concreteNodes() == null : "request concreteNodes shouldn't be set";
+        String[] nodesIds = clusterState.nodes().resolveNodes(request.nodesIds());
+        // TODO: version needs to be updated once 7.x backport is in place
+        request.setConcreteNodes(Arrays.stream(nodesIds).map(clusterState.nodes()::get)
+            .filter(node -> node.getVersion().onOrAfter(Version.V_8_0_0)).toArray(DiscoveryNode[]::new));
```
Added BWC handling.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestClearPrivilegesAction.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public List<Route> routes() {
+        return Collections.singletonList(new Route(POST, "/_security/privilege/{application}/_clear_cache"));
```
Do we need a `ReplacedRoute` here similar to the one in clear role cache action? i.e.:
https://github.com/elastic/elasticsearch/blob/022d3d7dd67fd882981f493a3228552b8a198a71/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/role/RestClearRolesCacheAction.java#L36-L37
My understanding is that the ReplacedRoute was needed because of xpack migration into main repo and is _not_ necessary for new APIs. So I didn't add it here.

ywangd committed:
Add working rolling upgrade mixed-cluster tests for privilege cache clearing

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

ywangd commented:
* Add [mixed-cluster test](https://github.com/elastic/elasticsearch/pull/55836/files#diff-cd629a343bf151f7a9dd1a23fbdb9972)

ywangd committed:
Remove bwc handling code and simply report no handler

ywangd committed:
Checkstyle

ywangd committed:
Clear privilege cache on security index changes

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -83,47 +97,95 @@
+
+    private void innerGetPrivileges(Collection<String> applications,
+        ActionListener<Collection<ApplicationPrivilegeDescriptor>> listener) {
+        assert Objects.requireNonNull(applications).isEmpty() == false;
```
That's a weird mix of `assert` and `require`.
```suggestion
        assert applications != null && applications.size() > 0 : "Application names are required (found " + applications + ")";
```

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -133,25 +195,75 @@ public void getPrivileges(Collection<String> applications, Collection<String> na
+        return descriptors.stream().filter(d -> privilegeNames.contains(d.getName())).collect(Collectors.toUnmodifiableSet());
     }
 
     private QueryBuilder getApplicationNameQuery(Collection<String> applications) {
```
Nit: I think the order of the methods in this file is a bit random.
This probably should be after `innerGetPrivileges` since it's an implementation detail of that method.

ywangd replied:
Not sure If I understand correctly, but this method (L269) is already after `innerGetPrivileges` (L173). 

In terms of overall method ordering, I have seen two preferences: 1) order by visibility with public first and private last; 2) group by related functions with public methods followed by their supporting private methods. 
My impression is the ES code base loosely follows option 2. This file is example of it. Do we have a recommended style?

tvernum replied:
Sorry, I was a bit loose with my terminology.
`getApplicationNameQuery` is _after_ `innerGetPrivileges`, but not _immediately after_.

I see 
- `innerGetPrivileges`
- `onSecurityIndexStateChange`
- `invalidate`
- `invalidateAll`
- `cachedDescriptorsForApplicationNames`
- `filterDescriptorsForPrivilegeNames`
- `getApplicationNameQuery`

I can't see any pattern to that order.

ywangd replied:
Thanks for the elaboration. I tried to organise them by following option 2. Private methods are listed after the public methods that are the callers. The result is as follows:

```java
public void getPrivileges
private void innerGetPrivileges
private QueryBuilder getApplicationNameQuery
private ApplicationPrivilegeDescriptor buildPrivilege
private Set<ApplicationPrivilegeDescriptor> cachedDescriptorsForApplicationNames
private Collection<ApplicationPrivilegeDescriptor> filterDescriptorsForPrivilegeNames

public void putPrivileges
private void innerPutPrivilege

public void deletePrivileges
private <T> void clearCaches
private static Tuple<String, String> nameFromDocId
private static String toDocId

public void onSecurityIndexStateChange
public void invalidate
public void invalidateAll

private static boolean isEmpty
```

Does above ordering look reasonable to you?

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -133,25 +195,75 @@ public void getPrivileges(Collection<String> applications, Collection<String> na
+                        new ParameterizedMessage("Searching for [{}] privileges with query [{}]",
+                            applications, Strings.toString(query)));
                     request.indicesOptions().ignoreUnavailable();
+                    // TODO: not parsing source of cached entries?
```
I don't think we need this TODO do we?
I think in this context it's simpler to parse everything than try and track doc changes (because we trust the cache if it's populated, we only go to the index when the cache is missing something).

ywangd replied:
You are right. I wrote the Todo when I was trying out a few optimisation ideas, e.g.: 1) only fetch things that are missing in cache; 2) fetch everything but only parsing what is missing in cache;

But later we decided to simplify things by always fetching everything if we have to hit the index and repopulate the cache to make sure it is always up to date. So this todo is no longer necessary.

tvernum reviewed (commented):
I still need to review the tests, but it's late and I don't know when I'll get back to do another pass.

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java:
```diff
@@ -0,0 +1,303 @@
+        final List<ApplicationPrivilegeDescriptor> applicationPrivilegeDescriptors = Arrays.asList(
+            new ApplicationPrivilegeDescriptor("app-1", "read", Set.of("a:b:c", "x:y:z"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-1", "write", Set.of("a:b:c", "x:y:z"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-1", "admin", Set.of("a:b:c", "x:y:z"), emptyMap()),
```
Should the action names be different? It seems hard to check that the right value is cached when they're all the same.

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java:
```diff
@@ -0,0 +1,303 @@
+        assertEquals("write", privileges[0].getName());
+
+        // A hacky way to test cache is populated and used by deleting the backing documents.
+        // The test will fail if the cache is not in place
```
I think this is a sign that this test shouldn't be an integration test like this. If it weren't quite so integrated you would be able to spy on the client used by the privilege store.

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java:
```diff
@@ -0,0 +1,303 @@
+        assertEquals(6, putPrivilegesResponse.created().values().stream().mapToInt(List::size).sum());
+    }
+
+    public void testGetPrivileges() {
```
```suggestion
    public void testGetPrivilegesUsesCache() {
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java:
```diff
@@ -0,0 +1,303 @@
+        assertEquals(0, privileges.length);
+    }
+
+    public void testWildcard() {
```
Something like
```suggestion
    public void testPopulationOfCacheWhenLoadingPrivilegesForAllApplications() {
```
?

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreTests.java:
```diff
@@ -114,42 +118,27 @@ public void cleanup() {
-        assertThat(getPrivilege, equalTo(sourcePrivilege));
-    }
 
-    public void testGetMissingPrivilege() throws Exception {
```
Why did this test get removed?
Isn't that still a valid scenario?

ywangd replied:
The method under tests, `NativePrivilegeStore#getPrivilege`, is removed. So there test is no longer necessary.

`NativePrivilegeStore#getPrivilege` is removed because the code now always fetch all privileges for an application for caching purpose.

tvernum replied:
Do we have a test somewhere for 
```
public void getPrivileges(Collection<String> applications, 
                          Collection<String> names,
                          ActionListener<Collection<ApplicationPrivilegeDescriptor>> listener)
```

when the application exists but none of the privilege names do? That feels like something worth testing.

ywangd replied:
I added back a modified version of `testGetMissingPrivilege`. But then I think it is not really necessary. Please let me explain:

The code now always issue a search request, from end result's perspective, the difference really is just how many SearchHit is returned. It could be 0, 1 or N. What you described here is a case of `0`. There are other cases of `0` hits, e.g. non-existent application, but they are not different in terms of search response.

Search response is turned into a collection of `ApplicationPrivilegeDescriptor` by `buildPrivilege` and `ScrollHelper#fetchAllByEntity`. What we do want test is `buildPrivilege` since it is part of the class and it is only relevant when SearchHit is greater than `0`, because when there is no SearchHit, it won't get executed at all and we probably can trust `ScrollHelper` to do its job by returning an empty list.

I guess this is probably why both [`testGetPrivilegesByWildcardApplicationName`](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreTests.java#L182) and [`testGetPrivilegesByStarApplicationName`](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreTests.java#L202) did not bother to test the `future` results (since they are just empty lists). And one could say they already suffice to be tests for `0` SearchHit.

With above being said, I decided to keep `testGetMissingPrivilege` since it is already added and an extra test is not a bad thing.

ywangd commented on x-pack/docs/en/rest-api/security/clear-privileges-cache.asciidoc:
```diff
@@ -0,0 +1,42 @@
+++++
+
+Evicts privileges from the native application privilege cache.
+
```
Yes makes sense. I'll add "The cache is also automatically cleared for applications that have their privileges updated."

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java:
```diff
@@ -410,6 +410,7 @@ public static void buildRoleFromDescriptors(Collection<RoleDescriptor> roleDescr
             final Set<String> applicationPrivilegeNames = applicationPrivilegesMap.values().stream()
                     .flatMap(Collection::stream)
                     .collect(Collectors.toSet());
+            // Role itself is cached, so skipping caching for application privileges
```
Good catch! This comment is obsolete and should be removed. It is for the initial iteration where I tried to separate this usages from others.

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

Co-authored-by: Tim Vernum <tim@adjective.org>

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

Co-authored-by: Tim Vernum <tim@adjective.org>

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/ClearPrivilegesCacheRequest.java:
```diff
@@ -0,0 +1,86 @@
+        return applicationNames;
+    }
+
+    public boolean isClearRolesCache() {
```
You are right. I'll change it to `clearRolesCache`, which feels more consistent with overall naming style of request classes.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -83,47 +97,95 @@
+
+        // TODO: We should have a way to express true Zero applications
+        final Set<String> applicationNamesCacheKey = (isEmpty(applications) || applications.contains("*")) ?
+            Set.of("*") : new LinkedHashSet<>(applications);
```
hmm I have no idea now. It was previously `Set.copyOf`, but somehow I changed it to `LinkedHashSet` last time I made the change to "always fetch for all requested applications if we have to hit the index". Maybe I was trying to experimenting something. Cannot remember ... I'll change it back to `Set.copyOf`.

ywangd committed:
Update x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

Co-authored-by: Tim Vernum <tim@adjective.org>

ywangd committed:
Update x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java

Co-authored-by: Tim Vernum <tim@adjective.org>

ywangd committed:
Update x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java

Co-authored-by: Tim Vernum <tim@adjective.org>

ywangd committed:
Address feedback

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

ywangd committed:
Address feedback

ywangd committed:
Fix docs

ywangd commented:
Thanks for the feedbacks. I addressed most of them. Still need work on two more items:
1. Check if the single node integration tests should be replaced by using spy object in EsTestCase.
2. Check wether we should improve the current concurreny control with `AtomicLong numInvalidation` by ReadWriteLock or something along this line.

I'll get to them in the next pass.

tvernum commented:
> Check if the single node integration tests should be replaced by using spy object in EsTestCase.

If we've got working tests, then I'm OK with them, it just feels like we ended up with a very procedural test due to the decision to use an integration test for it.

ywangd commented:
> If we've got working tests, then I'm OK with them, it just feels like we ended up with a very procedural test due to the decision to use an integration test for it.
>
It's been a while since I wrote the tests. Now I remember these intergation tests are in fact extra efforts. There are unit tests for caching behaviour, please see methods named as `NativePrivilegeStoreTests#testGetPrivilegesCacheXxxxx`. I was trying to be sure that the cache would work in a real node environment, hence the single node integration tests. 

As for the `numInvalidation` check, I don't think there is a perfect solution, not even with a ReadWriteLock. The purpose of the `numInvalidation` check is to avoid caching stale result. So the exact scenario it tries to help is like the follows:
1. Before the search to security index, take a record of the `numInvalidation`.
2. A search result is then performed against the security index for application privileges
3. Right after the search is complete, the application privileges gets updated. 
4. A cache invalidation is triggered by the update and increment `numInvalidation`.
5. Before caching the search response, check `numInvalidation` again. If the number is changed, skip caching because this means the cache is invalidated between search completion time and now.

The above workflow is inherently impossible to synchronise. Because we cannot hold back index update just because we have a search response to use. My ToDo note was "the cache could still get invalidated immediately after `numInvalidation` is checked in step 5". So we end up still caching stale results. 

There are also other possibilities, e.g. the request for caching invalidation is taking long so that even the search response is effectively stale, but `numInvalidation` does not change in time to reflect it. So we would end up cache stale results again. I think this scenario basically makes ReadWriteLock not useful: if the lock is added around the cache, preventing caching itself from updating or invalidating does not hold back the index changes.

I think what we can do here is best efforts. The current way of handling this thus looks sufficiently good to me. I think what I will do is actually to remove my ToDo comment.

ywangd committed:
Address feedback

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

ywangd committed:
Remove todo as it is pointless

ywangd commented:
@elasticmachine run elasticsearch-ci/1

ywangd committed:
Address feedback for readwritelock and cache invalidation

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

ywangd commented:
After discussion with @tvernum, it is agreed that a ReadWriteLock is necessary to achieve maximum correctness. Indeed we cannot completely avoid caching stale result. But we can ensure if stale result is cached, it will be invalidated as soon as possible and this is where the ReadWriteLock comes in. It works as the follows:

* Before caching the result, acquire a read lock
* Before invalidating the cache, acquire a write lock

The acquisition of a read lock ensures any invalidation requests will be held off until the current result is cached. If the current reseult is stale, e.g. because index is updated while it is being cached, the locking mechanism guarantees that invalidation will happen after current caching action finishes. In another word, the time window for a stale result to stay in the cache is as small as possible.

The code is updated accordingly.

ywangd committed:
add tests for caching behaviour

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

ywangd committed:
Add TTL as a safety net. Also consolidate cache size to a single setting

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache

ywangd commented:
Two more updates:
* Add a ttl (default 24h) to the caches. This setting acts as a safety net in rare cases where things go out of sync and we ensure out-of-sync info will be cleared out eventually.
* Use a single the cache size setting for both caches because the two layer caching is an implementation detail and does not feel right to enforce users to be aware of it.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -346,4 +419,55 @@ private static String toDocId(String application, String name) {
+            // Always completely invalidate application names cache due to wildcard
+            applicationNamesCache.invalidateAll();
+            uniqueNames.forEach(descriptorsCache::invalidate);
+        }
```
I think we can release the lock immediately after incrementing `numInvalidation`. It will mean that, in theory, we could invalidate things that we don't need to, but would mean holding a lock for less time.

Did you consider the trade-off of how long to lock for _vs_ perfect cache accuracy?

If we keep the lock around all the invalidation process, then I think the calculation of `uniqueNames` should be before the lock is acquired.

ywangd replied:
Thanks @tvernum. Your comment is very insightful. You are right we could minimize the locking time in this case. The possibility of invalidating more than necessary should be very low. The `getPrivileges` thread needs to read the incremented value of `numInvalidation` and perform a search query and all these have to complete before the cache is invalidated. The chance should be extremely low. I have updated the code to minimize the locking for both `invalidate` and `invalidateAll`. 

A more possible scenario is "unnecessarily skipping put result in cache". We only cache when `numInvalidation` does not change. But this value changes for both partially and full cache invalidation. In the case of partial invalidation, the things get invalidated may not be relevant to the things that we want to cache. But the code would just skip caching them regardlessly for simplicity. This however has nothing to do with the locking, i.e. the same situation exists before we added the locking. Overall, combined with how Kibana behaves, I think this is an acceptable trade-off because:
* The chance is still low 
* We always fully invalidate `applicationNamesCache` even for partially invalidation. And we decide to keep it this way for simplicity.

Also moved `descriptorsCache != null` check before the read lock, so we do not just lock and find out there is no cache to use. It is an edge case optimization but anyway it is easy to add.

tvernum reviewed (approved):
LGTM, but I think we can be a bit smarter with the locking.

ywangd committed:
Address feedback about lock duration

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-54317-app-privilege-cache
